### PR TITLE
system tests for pre-refactor code

### PR DESCRIPTION
**Pull request by rajnutakki**
#17 
This is a system test for the old code, one with a fake LLM call and another with real LLM calls.
Both use the synthetic data problem but with a slightly modified `spec.py` which lives in the `tests/system` directory.  
I made minimal changes to `run.py` and `hypothesis_engine.py` to allow the use of the fake LLM.

At the moment they test only a small number of things, which are read from the `program_generation_log.jsonl`.  summarized below:
**Fake LLM test (`test_run_fake.py`):**
* Tests that it runs
* Tests it outputs the `program_generation_log.jsonl`
* Checks the total number of programs generated
* Checks the number of programs for which the test loss (validation loss) is calculated, i.e the number of surviving programs at the end.
* Checks the correct winner is identified (we know this from the design of the fake LLM).
* Checks the winner has a predetermined train and test loss (this is just the value I got from a previous run, so assumes the code gives correct values for these in its current state).

**Real LLM test (`test_run.py`):**
* Tests that it runs
* Tests it outputs the `program_generation_log.jsonl`
* Checks the total number of programs generated
* Checks the number of programs for which the test loss (validation loss) is calculated, i.e the number of surviving programs at the end.

I think it's sufficient to check this small number of things, more precise checking of correct behaviour can be done in unit tests. But let me know if you think something should be added.

### CODE REVIEW

File: tests/system/spec.py [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+    n_trials: int = 2000,
+    noise_std: float = 0.1,
+) -> list[list[dict[str, np.ndarray]]]:
+    """
+    Simulate synthetic single-input regression data.
+
+    Data-generating process:
+    y = (a*x^2 + b*x + c) * sin(k*x + phi_0) + epsilon,
+    where epsilon ~ N(0, noise_std^2), and each sample has its own parameters.
+    For test purposes, use zero noise and generate a,b,c,k,phi_0 deterministically based on n_samples
```
Comment (dabin-k) [inline tests/system/spec.py:51]: Do we really need this separate spec.py? IF we are using a fixed random seed to generate noise, the error should still be fixed and verifiable with hard coded values?
Perhaps there is something else that's different between the two specs which I've missed?

File: tests/system/config2.yaml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  n_samples: 4
+  n_trials: 100
+  noise_std: 0.0
+
+
+# ========================
+# Experiment Runtime
+# ========================
+
+# Runtime overrides for this task.
```
Comment (dabin-k) [inline tests/system/config2.yaml]: Do we gain much by keeping two config yamls? Can we just use the same one for both tests?

**Review by dabin-k — COMMENTED** [review]
I think the expect test with fake LLM is good and checks the deterministic features we want to check. 
But I don't think the LLM using expect test (test_run) achieves anything extra that isn't already captured by either the new expect_test or the test_mode, so we should consider removing it.

Another thing we can check more explicitly is the LLM calls/API client. A couple of things I've thought of so far 
1. models available - Use client.aio.models.list() to list available models and check that all models used in all projects are still available.
2. super lightweight async client.aio.models.generate_content check - e.g. a single ping with a short prompt like - "Reply with exactly : ping"
But maybe this is beyond the scope of this feature. We can make this feature separately.

File: tests/system/test_run_fake.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+def test_correct_winner(fake_run):
+    """ Verify the final winning program is the (0,0,2) program, which is the ProgramSolution without offset"""
+    programs = load_programs(fake_run)
+    winning_programs = [p for p in programs if p.get("is_winner")]
+    assert len(winning_programs) == 1, "Expected exactly 1 winning program"
+    winning_birth_details = (winning_programs[0]["iteration_number"], winning_programs[0]["birth_island"], winning_programs[0]["batch_index"])
+    assert winning_birth_details == (0,0,2), f"Expected winning program born at (iteration 0, island 0, batch index 2), found {winning_birth_details}"
+
+def test_winning_loss(fake_run):
```
Comment (dabin-k) [inline tests/system/test_run_fake.py:92]: We could make this check more granular and separately check the initial loss after parameter_estimator and loss after gradient descent to expicitly check the gradient descent process. If it requires more than a few lines of code, maybe not worth it.

File: tests/system/test_run_fake.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    """ Verify the winning program has anticipated train and test loss"""
+    programs = load_programs(fake_run)
+    winning_program = [p for p in programs if p.get("is_winner")][0]
+    assert round(winning_program["train_loss"],3) == 0.056, f"Expected winning program train_loss 0.056, found {winning_program['train_loss']:.3f}"
+    assert round(winning_program["test_loss"],3) == 0.141, f"Expected winning program test_loss 0.141, found {winning_program['test_loss']:.3f}"
```
Comment (dabin-k) [inline tests/system/test_run_fake.py]: We could also test things like timeouts for parameter_estimators and whether the the codebase accurately times out and allocates infinite loss. But this is perhaps also beyond the scope of this feature..

File: tests/system/spec.py (hunk 1 above)
Comment (rajnutakki) [reply tests/system/spec.py:51]: I agree the output will be fixed but I found it helpful to get exactly the behaviour I want.
I removed the use of random numbers to generate things, which just meant that I could fix target function parameters so that the program I want is the winner without much parameter optimisation. 

I also think it's a good idea to separate the test and project spec. 
Someone may want to modify the synthetic data project spec in their code, but if they did it wrong it doesn't mean the system test should fail, i.e the code works fine, just their project spec is wrong.
I would have separate tests for the projects.

Let me know if you agree?

File: tests/system/config2.yaml (hunk 2 above)
Comment (rajnutakki) [reply tests/system/config2.yaml]: This one has less parameter optimisation etc.. to make runtime quicker, since we don't actually care about final values. 
But I agree (see comment below) that it is essentially the same as test_mode.
@reillytilbury will test_mode also be in the refactor? If so I can just make this run with real LLM calls a test_mode run, as it will still be good to have this as a pytest so that we can automatically check the test_mode runs without error.

File: tests/system/test_run_fake.py (hunk 4 above)
Comment (rajnutakki) [reply tests/system/test_run_fake.py]: I think this is beyond the scope. These would be good unit tests for the refactored code.

File: tests/system/test_run_fake.py (hunk 3 above)
Comment (rajnutakki) [reply tests/system/test_run_fake.py:92]: Done. See [7858f5d](https://github.com/reillytilbury/EDGAR/pull/18/changes/7858f5dae7a8cd1ff5a032d8d394bddd680c5158)

**Comment by rajnutakki** [comment]
> I think the expect test with fake LLM is good and checks the deterministic features we want to check. But I don't think the LLM using expect test (test_run) achieves anything extra that isn't already captured by either the new expect_test or the test_mode, so we should consider removing it.
> 
> Another thing we can check more explicitly is the LLM calls/API client. A couple of things I've thought of so far
> 
>     1. models available - Use client.aio.models.list() to list available models and check that all models used in all projects are still available.
> 
>     2. super lightweight async client.aio.models.generate_content check - e.g. a single ping with a short prompt like - "Reply with exactly : ping"
>        But maybe this is beyond the scope of this feature. We can make this feature separately.

Thanks for your comments. 

- I agree that test_run basically does the same as test_mode, just we should have this formalized into a test so we can check it runs automatically. How to do this depends on whether the refactored code has a test_mode or not (?).
- I added in a check of the `initial_loss` to the test of the loss of the winning program for the fake LLM call ( [7858f5d](https://github.com/reillytilbury/EDGAR/pull/18/changes/7858f5dae7a8cd1ff5a032d8d394bddd680c5158)), so loss is checked before and after gradient descent
- I think other checks like these LLM calls are good unit tests which should be implemented for the refactored code, but not here.

In https://github.com/reillytilbury/EDGAR/commit/a4c90f43188fe8a4feea13f210d724762ce89aa2, I:
- Switched to a single config.yaml
-   `test_run` runs a test_mode run with real LLM calls.
- Added a program generated by the fake LLM which causes an error, and check I get None for its loss and that it's test_loss is not computed at the end (it didn't survive to the end).

File: tests/system/config2.yaml (hunk 2 above)
Comment (dabin-k) [reply tests/system/config2.yaml]: Could we not have fewer numbers of steps for gradient descent for both config.yaml? (that's what you mean by less parameter optimisation right?) 
I think I'd be happy to verify the initial loss with an updated loss value that corresponds to the changed n_steps if that's the problem

File: tests/system/spec.py (hunk 1 above)
Comment (dabin-k) [reply tests/system/spec.py:51]: Fair enough. Sounds good!

File: tests/system/config2.yaml (hunk 2 above)
Comment (rajnutakki) [reply tests/system/config2.yaml]: I've sorted this out now in https://github.com/reillytilbury/EDGAR/commit/a4c90f43188fe8a4feea13f210d724762ce89aa2.

I do fewer number of gradient descent steps for the fake LLM and now use test_mode = True to ensure I don't use many LLM calls for that test.
So `test_run` is just an automated run of `test_mode` that checks the output looks correct also.
